### PR TITLE
Define and implement `DocumentStore`

### PR DESCRIPTION
**Pull request by sandreae**
Makes use of https://github.com/p2panda/p2panda/pull/319

## `DocumentStore` trait definition
- note: we don't have custom storage structs here, we require the p2panda-rs types `DocumentView` and `Document`
```rust

/// Storage traits for documents and document views.
#[async_trait]
pub trait DocumentStore {
    /// Insert document view into storage.
    ///
    /// returns an error when a fatal storage error occurs.
    async fn insert_document_view(
        &self,
        document_view: &DocumentView,
        schema_id: &SchemaId,
    ) -> Result<(), DocumentStorageError>;

    /// Get a document view from storage by it's `DocumentViewId`.
    ///
    /// Returns a DocumentView or `None` if no view was found with this id. Returns
    /// an error if a fatal storage error occured.
    async fn get_document_view_by_id(
        &self,
        id: &DocumentViewId,
    ) -> Result<Option<DocumentView>, DocumentStorageError>;

    /// Insert a document into storage.
    ///
    /// Inserts a document into storage and should retain a pointer to it's most recent
    /// document view. Returns an error if a fatal storage error occured.
    async fn insert_document(&self, document: &Document) -> Result<(), DocumentStorageError>;

    /// Get the lates document view for a document identified by it's `DocumentId`.
    ///
    /// Returns a type implementing `AsDocumentView` wrapped in an `Option`, returns
    /// `None` if no view was found with this document. Returns an error if a fatal storage error
    /// occured.
    ///
    /// Note: if no view for this document was found, it might have been deleted.
    async fn get_document_by_id(
        &self,
        id: &DocumentId,
    ) -> Result<Option<DocumentView>, DocumentStorageError>;

    /// Get the most recent view for all documents which follow the passed schema.
    ///
    /// Returns a vector of `DocumentView`, or an empty vector if none were found. Returns
    /// an error when a fatal storage error occured.  
    async fn get_documents_by_schema(
        &self,
        schema_id: &SchemaId,
    ) -> Result<Vec<DocumentView>, DocumentStorageError>;
}
```

## DB structure
```
CREATE TABLE IF NOT EXISTS document_view_fields (
    document_view_id            TEXT            NOT NULL,
    operation_id                VARCHAR(64)     NOT NULL,
    name                        VARCHAR(128)    NOT NULL,
    FOREIGN KEY(operation_id)   REFERENCES      operations_v1(operation_id)
);

CREATE INDEX idx_document_view_fields ON document_view_fields (document_view_id, operation_id, name);

CREATE TABLE IF NOT EXISTS document_views (
    document_view_id            TEXT            NOT NULL UNIQUE,
    schema_id                   TEXT            NOT NULL,
    PRIMARY KEY (document_view_id)
);

CREATE TABLE IF NOT EXISTS documents (
    document_id                 VARCHAR(64)     NOT NULL UNIQUE,
    document_view_id            TEXT            NOT NULL,
    schema_id                   TEXT            NOT NULL,
    is_deleted                  BOOLEAN         NOT NULL,
    PRIMARY KEY (document_id)
);

```

## 📋 Checklist

- [x] Add tests that cover your changes
- [x] Add this PR to the _Unreleased_ section in `CHANGELOG.md`
- [x] Link this PR to any issues it closes
- [x] New files contain a SPDX license header

### CODE REVIEW

File: aquadoggo/src/db/stores/document.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    use p2panda_rs::operation::{AsOperation, OperationId};
+    use p2panda_rs::schema::SchemaId;
+    use p2panda_rs::storage_provider::traits::{AsStorageEntry, EntryStore};
+    use p2panda_rs::test_utils::constants::{DEFAULT_HASH, DEFAULT_PRIVATE_KEY, TEST_SCHEMA_ID};
+
+    use crate::db::models::entry::EntryRow;
+    use crate::db::stores::document::{DocumentStore, StorageDocumentView};
+    use crate::db::stores::test_utils::{test_create_operation, test_db};
+
+    fn entries_to_document_views(entries: &Vec<EntryRow>) -> Vec<StorageDocumentView> {
```
Comment (sandreae) [inline aquadoggo/src/db/stores/document.rs:177]: I think i should be using the new mocking features on the stores for these tests...

File: aquadoggo/src/db/traits/document.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        document_view: &StorageDocumentView,
+        schema_id: &SchemaId,
+    ) -> Result<(), DocumentStorageError>;
+
+    async fn get_document_view_by_id(
+        &self,
+        id: &DocumentViewId,
+    ) -> Result<StorageDocumentView, DocumentStorageError>;
+
+    async fn get_document_views_by_schema(
```
Comment (sandreae) [inline aquadoggo/src/db/traits/document.rs:62]: I don't think we actually want this method.... My mistake!

**Comment by sandreae** [comment]
I think we can have a cool method which looks like:

```
insert_get_document_view_by_id() -> Result<Option<DocumentView>, Error>;
```

Which would:
- Find the document which contains this document view
- Collect all of the operations which exist for this document
- Resolve the document graph up until the requested document id
- Store this document view in the db and return it

This is interesting because then we can retrieve document views which haven't been explicitly inserted yet just by their id. Could be useful in materialisation, when we know a document view id, but don't know if it's been materialised yet (the operations required may exist, but it was never materialised into a document view).

**Comment by sandreae** [comment]
needs rebasing onto `development` once we merge #103

**Comment by adzialocha** [comment]
Upsi, shouldn't have deleted that branch from the merged PR, sorry! 😬😵

**Comment by sandreae** [comment]
Oh, sorry, not sure what happened. This PR shouldn't be closed though....

**Comment by adzialocha** [comment]
> Ah, cos it was targeting the deleted branch. Cool, I rebased and pushed it anyway 👍

Yeah, it's GitHub doing it automatically. Realized that we (me) could stop always deleting branches after merges (it's a habit) to prevent it.

**Comment by cafca** [comment]
We could clean them up every once in a while from github's branch oiverview page

**Pull request closed without merging**